src/java.desktop/unix/classes/sun/awt/UNIXToolkit.java

```diff
@@ -240,6 +244,71 @@ protected Object lazilyLoadGTKIcon(String longname) {
         return img;
     }
 
+    private static volatile Boolean shouldDisableSystemTray = null;
+
+    /**
+     * There is an issue displaying the xembed icons in appIndicators
+     * area with certain Gnome Shell versions.
+     * To avoid any loss of quality of service, we are disabling
+     * SystemTray support in such cases.
+     *
+     * @return true if system tray should be disabled
+     */
+    public boolean shouldDisableSystemTray() {
+        Boolean result = shouldDisableSystemTray;
+        if (result == null) {
+            synchronized (GTK_LOCK) {
+                result = shouldDisableSystemTray;
+                if (result == null) {
+                    if ("gnome".equals(getDesktop())) {
+                        @SuppressWarnings("removal")
+                        Integer gnomeShellMajorVersion =
+                                AccessController
+                                        .doPrivileged((PrivilegedAction<Integer>)
+                                                this::getGnomeShellMajorVersion);
+
+                        if (gnomeShellMajorVersion == null
+                                || gnomeShellMajorVersion < 45) {
+
+                            return shouldDisableSystemTray = true;
+                        }
+                    }
+                    shouldDisableSystemTray = result = false;
+                }
+            }
+        }
+        return result;
+    }
+
+    private Integer getGnomeShellMajorVersion() {
+        try {
+            Process process =
+                new ProcessBuilder("/usr/bin/gnome-shell", "--version")
```

Comment:
How it will affect performance?

Reply:
Well, clearly the exec will take some non-zero time, but this part of initialising the SystemTray only happens once and (arm-waving alert) 99% of apps don't do it at all, so I don't see it as a major problem and it needs to be done.
There's a plan to re-do it with an API call later but that is more complex code than we have time for now, so it will be fine as a temporary solution.

Reply:
According to my tests, the call to `getGnomeShellMajorVersion()` took about 25-35 ms on average, but it is a one-time task.

Reply:
> (arm-waving alert) 99% of apps don't do it at all,

That may not be so arm-wavy actually considering this has been a problem for a while and we haven't exactly been inundated with complaints, especially since on RH & (I think) Fedora distros SystemTray isn't supported anyway unless you install gnome-shell-extensions.

But hmm .. there's a thought - so long as available returns false anyway (via the existing code) do we even need to make this call ? Might need to be careful about when/whether the selectionListener is added.
Alexander can you look at that to see if we can just not bother with the exec if it isn't going to be needed anyway ?

Reply:
If we do that we could be up to 99.5% of cases aren't affected.

Reply:
And in fact we need to be VERY careful that we don't turn the existing code reporting "available == false" into "true" due to us simply detecting Gnome >= 45 .. that would be unfortunate.

Reply:
Latest comment from Phil related to not disabling SystemTray for Gnome>=45 needs to be taken care.
Also do we need to worry about this issue being regression from some Gnome version like 3.38.0 and we need to disable SystemTray only between 3.38.0 & < 45 Or we are okay with disabling it for all version less than 45?

Reply:
A fair question. My understanding of how Gnome vs Gnome Shell vs GTK version naming interact here is definitely not as clear as I'd like but I understand anything that is Gnome 3.X or 40+ is affected, leaving only the even older systems out there as theoretically working. ie anything in at least the last 5-6 years is broken. I'd like Alexander to chime in here on this too.

Reply:
> Latest comment from Phil related to not disabling SystemTray for Gnome>=45 needs to be taken care.

With the current version string format, this should not happen unless the executable is somewhere else, environment variables are stripped, etc.

However, since this is a quick fix, a more reliable way to do this will be done along with  [JDK-8325914](https://bugs.openjdk.org/browse/JDK-8325914).


> Also do we need to worry about this issue being regression from some Gnome version like 3.38.0 and we need to disable SystemTray only between 3.38.0 & < 45 Or we are okay with disabling it for all version less than 45?

The 3.38.0 was just an example of a different numbering convention, I can reproduce the issue on the 3.28.3 on OL7.9. So the lower limit is not known yet.  So this is intentional, as an urgent fix.

Reply:
PS for context, one possibility strongly considered was to disable it EVERYWHERE, regardless of version.

Reply:
Yes i know 3.38.0 was just an example i used. I was coming from the perspective of "Are we okay with disabling SystemTray for all versions <45 irrespective of whether SystemTray might work in some earlier version?". With latest comments in PR i see that we are okay with disabling for all <45 since it is a temporary urgent fix.